Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -25,30 +25,68 @@
+void ScrollAnchoringController::invalidateAnchorBox()
 {
-    m_anchorElement = nullptr;
-    m_lastPositionForAnchorElement = FloatPoint();
+    m_anchorBox = nullptr;
+    m_lastPositionForAnchorBox = FloatSize();
 }
 
-void ScrollAnchoringController::selectAnchorElement()
+FloatSize ScrollAnchoringController::computeOffset(RenderBox* candidate)
 {
-    // TODO: implement
-    // perform Anchor Element Selection algorithm on dom elements within frame view's viewable region
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return candidate->absoluteBoundingBoxRect().location() - m_frameView.layoutViewportRect().location();
+}
+
+void ScrollAnchoringController::chooseAnchorBox()
+{
+    // TODO: fully implement anchor node selection algorithm in spec
+    // (specifically excluded subtrees and priority nodes)
+    // https://drafts.csswg.org/css-scroll-anchoring/#anchor-node-selection
+    auto scrollPosition = frameView().scrollPosition();
+    if (m_anchorBox || scrollPosition.isZero())
+        return;
+    
+    auto& document = *m_frameView.frame().document();
+    auto* documentElement = document.documentElement();
+    RenderBox* rootRenderer = documentElement ? documentElement->renderBox() : nullptr;
+    RenderBox* candidateAnchor = nullptr;
+    auto layoutViewport = m_frameView.layoutViewportRect();
+    
+    Vector<RenderBox*> boxList;
+    boxList.append(rootRenderer->firstChildBox());
+    while (!boxList.isEmpty()) {
+        auto box = boxList.takeLast();
+        if (box->style().overflowAnchor() == OverflowAnchor::None)
+            continue;
+        auto boxRect = box->absoluteBoundingBoxRect();
+        if (layoutViewport.intersects(boxRect))
+            candidateAnchor = box;
+        if (layoutViewport.contains(boxRect))
+            break;
+        if (!is<RenderBox>(box->firstChild()))
+            continue;
+        for (auto child = box->firstChildBox(); child; child = child->nextSiblingBox()) {
+            boxList.append(child);
+            if (!is<RenderBox>(child->nextSibling()))
+                break;
+        }
+    }
+    if (!candidateAnchor)
+        return;
+    m_anchorBox = candidateAnchor;
+    m_lastPositionForAnchorBox = computeOffset(m_anchorBox.get());
 }
 
 void ScrollAnchoringController::updateScrollPosition()
 {
-    // TODO: implement
-    // calling m_frame.scrollTo with different between m_anchorElement current position
-    // and m_lastPositionForAnchorElement
-    // should be called after suppression window (maybe Page::doAfterUpdateRendering())
+    if (!m_anchorBox)
+        return;
+    FloatSize adjustment = computeOffset(m_anchorBox.get()) - m_lastPositionForAnchorBox;
+    if (!adjustment.isZero())
+        m_frameView.setScrollPosition(frameView().scrollPosition() + IntPoint(adjustment.width(), adjustment.height()));
```

Comment:
Please add logging with `LOG(Scroll, ...)` here so we know when scroll anchoring is changing the scroll position.

Reply:
Fixed.